### DISCUSSION TITLE
Control over the order in which dependencies are updated?

### DISCUSSION THREAD

**Discussion by karfau** [renovatebot/renovate#10217] [Request Help] [answered]
**What would you like Renovate to be able to do?**

When enabling a limit to how many PRs/branches are created at once and enabling stability days, I have observed that often dependencies just pile up because the order of the dependencies that are trying to be updated just seems to be alphabetic, so the depenencies "earlier" in the alphabet can block/delay those "later in the alphabet". Which also means that all scoped packages come before any unscoped packes, which means that many very "old" and well known unscoped packages do not get updated because in case of e.g. a repo using @types/... packages, there are new type updates nearly every day.

Wouldn't it make sense to have an option for other ways of ordering the updates?

The way that would make most sense to me, especially when stabilityDays check is enabled (and a limit is enabled?), would be to first update the dependency with the "oldest" update, since it's less likely to block others, since the stabilityDays already passed or will pass first.

Or at least to prefer updates that would already pass the stabilityDays check before those that doesn't?

Maybe just picking at random would also be a valid way order updates?

**Did you already have any implementation ideas?**

No.

**Comment by rarkins** [comment, marked as answer]
Updates are prioritized first by updateType, e.g. with digest coming before patch, patch before minor, and minor before major. In general this is definitely the best order in terms of ability to merge.

I'm not aware of us doing any sorting alphabetically, but if your dependencies are sorted within your package files then that order is likely maintained.

There is an option to prioritize: https://docs.renovatebot.com/configuration-options/#prpriority

You could instead set[`internalChecksFilter=strict`](https://docs.renovatebot.com/configuration-options/#internalchecksfilter) so that pending updates are suppressed and non-pending would be raised first. I think this likely solves the majority of your problem.

Random is not a feasible idea as it would introduce unpredictability and noise if you can't assure the same "random" each run.

**Reply by karfau** [reply]
Thank you this works like a charm. Not sure why I can not mark you reply as "the answer".

**Reply by rarkins** [reply]
It was classified as "Ideas & General Discussion" which is an open format without the concept of there being an "answer". But I reclassified it just now and selected it per your wishes

**Comment by jacobrask** [comment]
I’d still want the features initially suggested. 

We have a huge backlog of dependencies to update, and want to prioritize the ones where we’re the most behind. It’s a monorepo with a number of important apps so we can’t reset the lockfile or anything like that, it’s way too much risk.

Currently we’ll update the first dependencies in our list and keep updating the patch versions for those but never get to the more important ones.

Ideally we could run a script like [libyear](https://github.com/jdanil/libyear) and use the drift value to prioritize updates. Is that possible already somehow or would it require changes to Renovate?

**Reply by rarkins** [reply]
Won't that essentially mean you want updates in the *opposite* order to which Renovate does today? Biggest impact first, i.e. major updates, then minor, etc?

**Reply by jacobrask** [reply]
That might be the case yes - because otherwise we'll never get to the major versions. We could also potentially prioritise only between major versions with the drift value of libyear.

**Reply by jacobrask** [reply]
We have a monorepo where some, but far from all, dependencies are pinned to the same version across the repo. Each application in the repo gets separate Renovate PRs through packageRules and groups.

A separate `prConcurrentLimit` per group would be another way of distributing the PRs - we could set max PRs open per application instead. Right now the app starting with `c` gets a bunch of repeated patch updates but the app starting with `t` gets no updates.

I feel that the spirit of libyear is similar to allowing a higher priority for security updates though. Even though there is no known open vulnerability, it's a higher risk for me to be on a very outdated version of one dependency than to be 2 patch versions behind on something that releases weekly.

**Reply by karfau** [reply]
Did you consider letting renovate create and maintain the Dependency Dashboard issue and manually checking boxes there in the order that your prefer?

**Reply by jacobrask** [reply]
We've done that to prioritize some dependencies, but it relies on continuous manual work basically every day for a long time.

If I could do it at one point - mark a large number of dependencies and those would be created over the cause of a few weeks but staying below the pr limit - that might be feasible. But if we could run a custom command like libyear, perhaps when fetching changelogs, and then use that as template variables to decide prio and groups, that would be ultimate flexibility that would work both now in the initial phase, and later when our queue is shorter.

**Reply by rarkins** [reply]
I think that you could achieve most of what you want using packageRules and prPriority settings. I.e. give major updates highest priority, then minor, then patch

**Reply by flin-8** [reply]
> I think that you could achieve most of what you want using packageRules and prPriority settings. I.e. give major updates highest priority, then minor, then patch

I'm hoping to get major updates highest priority, then minor, then patch

@rarkins could you elaborate on how using packageRules and prPriority settings can achieve this? doesn't prPriority work on packages? Is it creating 3 package rules (major, minor and patch) for each package, then setting prPriority on them?

**Reply by rarkins** [reply]
Yes that's how you'd do it.
